### PR TITLE
Implement metadata extraction from composer.json files in proxy

### LINKED ISSUE #3: Implement metadata extraction from composer.json files in proxy
Composer projects contain metadata in their `composer.json` files (https://getcomposer.org/doc/04-schema.md). In other formats we would extract this information when proxying or hosting an artifact, and the information would be included in the format attributes and search attributes.

We are currently not extracting any of this information and should write a general-purpose extractor for this information and store the extracted information. For the proxy implementation this operation should be relatively straightforward since we will only be extracting content from the downloaded source archive rather than a .phar file (which is somewhat different from a .zip or .tar). We'll worry about handling .phar files later if we have to.

Note that this JSON is different than the provider JSON that constitutes an index for purposes of Composer.

### PR DESCRIPTION
**Pull request by fjmilens3**
Resolves #3 by extracting the following `composer.json` fields as format attributes:

- name
- description
- version
- type
- keywords
- homepage
- time
- license
- authors
- support

Not perfect by any means but not that different than what we do elsewhere, and at least gets _something_ in place for some of the bigger gaps in the current implementation.

### CODE REVIEW

File: src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractor.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    }
+  }
+
+  /**
+   * Processes a single entry in the archive. If the entry is the composer.json then the attributes will be extracted.
+   * If not, the entry is skipped.
+   */
+  private boolean processEntry(final ArchiveInputStream stream,
+                               final ArchiveEntry entry,
+                               final NestedAttributesMap formatAttributes) throws IOException
```
Comment (doddi) [inline src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractor.java]: Question: Is there a reason you chose to pass the AttributesMap rather than returning a built AttributesMap to be set into the Asset?

File: src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractor.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   * Extracts author contact information (except for the role) into a collection of strings.
+   */
+  @VisibleForTesting
+  void extractAuthors(final Map<String, Object> contents,
+                      final NestedAttributesMap formatAttributes)
+  {
+    Object sourceValue = contents.get(AUTHORS);
+    if (sourceValue instanceof Collection) {
```
Comment (doddi) [inline src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractor.java]: Trivial: Should we log if the author information could not be found in the composer file

File: src/test/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractorTest.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+  }
+
+  @Test
+  public void extractInfoFromZipballWithoutJson() throws Exception {
+    try (InputStream in = getClass().getResourceAsStream("extractInfoFromZipballWithoutJson.zip")) {
+      when(tempBlob.getBlob()).thenReturn(blob);
+      when(blob.getInputStream()).thenReturn(in);
+
+      NestedAttributesMap attributesMap = new NestedAttributesMap("composer", new LinkedHashMap<>());
+      ComposerFormatAttributesExtractor underTest = new ComposerFormatAttributesExtractor();
```
Comment (doddi) [inline src/test/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractorTest.java]: This is used in most tests so could go in setup.

File: src/test/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractorTest.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+      underTest.extractFromZip(tempBlob, attributesMap);
+
+      assertThat(attributesMap.keys(), is(empty()));
+    }
+  }
+
+  @Test
+  public void extractStringsMissing() {
+    NestedAttributesMap destination = new NestedAttributesMap("backing", new LinkedHashMap<>());
```
Comment (doddi) [inline src/test/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractorTest.java]: Trivial: You could consider moving this to setup as it is used an a majority or cases (if you decide to use setup from the comment above)

File: src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractor.java (hunk 1 above)
Comment (fjmilens3) [reply src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractor.java]: > Question: Is there a reason you chose to pass the AttributesMap rather than returning a built AttributesMap to be set into the Asset?

Just that we were going to be setting all these into the formatAttributes portion of a NestedAttributesMap, and so you can't just "set" those values in (unless you're going to try and modify the backing map yourself, which seems to counter the intent behind the NestedAttributesMap in the first place)

I agree that it's inelegant to do it this way but it seemed no worse than returning a map and then having to iterate over all the keys to copy them into the destination formatAttributes NestedAttributesMap one by one, which is what we usually end up having to do in these situations. (Though if I were going to do this over, I might consider doing it that way since I do think the code ends up being more testable.)

File: src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractor.java (hunk 2 above)
Comment (fjmilens3) [reply src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractor.java]: > Trivial: Should we log if the author information could not be found in the composer file

I'm not that worried about a missing piece of information like this since it's not something germane to the actual functioning of the implementation; we don't validate each piece of metadata for the other formats, though I concede in this case we could make an argument for doing so because of the existence of a defined schema for the JSON. When we get to the hosted impl there will be some fields that we do care about (we cannot proceed without them, i.e. the version field) in general if there's a missing field.

File: src/test/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractorTest.java (hunk 3 above)
Comment (fjmilens3) [reply src/test/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractorTest.java]: > This is used in most tests so could go in setup.

Overall I've been leaning more toward just keeping the declarations as close as possible to where they're used, but in this cases they do have a lot of commonality. Done in https://github.com/sonatype-nexus-community/nexus-repository-composer/pull/11/commits/97b0aa140b325bb5be24906969c0f9a5a937c843 and https://github.com/sonatype-nexus-community/nexus-repository-composer/pull/11/commits/0e02ec7317a08e6e1c67c469b4990e14c3124684.

Comment (jlstephens89) [reply src/test/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractorTest.java]: This is more my own musings/findings rather than direction for anyone else but I've found that defining the "happy path" in the setup method but then keeping the setup for the deviant cases within the tests reduces duplication and makes the tests easier to read. It also makes the edge cases easier to spot because they're the ones with the "complicated" tests.

File: src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerContentFacetImpl.java [hunk 5]
```diff
@@ -190,6 +195,14 @@ protected Content doPutContent(final String path,
         false
     );
 
+    try {
+      asset.formatAttributes().clear();
```
Comment (jlstephens89) [inline src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerContentFacetImpl.java]: Note: This behaviour is something a future modifier would have to be aware of if for some reason we needed to add any derived attributes to this map. I would be inclined to only remove the composer supplied attributes or alternatively move this into a class like ComposerAttributeUpdate to make the behaviour more obvious.
